[quant][fx][fix] Fused modules with object_type in qconfig

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60883
* #60653
* **#60779**
* #60378
* #60272
* #60034
* #59964

Summary: When we do fusion, we replace certain modules (such as Linear + ReLU) with fused versions (such as LinearReLU) by calling `_fuse_fx` in prepare_fx. However when we try to look up using the fused module type in qconfig_dict, we cannot find a match anymore since the qconfig dict contains the original module types. An example is here [N882873](https://fburl.com/anp/azenjx3v). 

So we will now update the qconfig_dict to include the fused modules mapping to the qconfigs used for the modules that make up the fused modules. If the modules are not mapped to the same qconfig, then we will raise an error.

Test Plan: `python test/test_quantization.py TestFuseFx.test_qconfig_fused_module`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29406941](https://our.internmc.facebook.com/intern/diff/D29406941)

angelayi committed:
[quant][fx][fix] Fused modules with object_type in qconfig

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx][fix] Fused modules with object_type in qconfig"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx][fix] Fused modules with object_type in qconfig"


Summary: When we do fusion, we replace certain modules (such as Linear + ReLU) with fused versions (such as LinearReLU) by calling `_fuse_fx` in prepare_fx. However when we try to look up using the fused module type in qconfig_dict, we cannot find a match anymore since the qconfig dict contains the original module types. An example is here [N882873](https://fburl.com/anp/azenjx3v). 

So we will now update the qconfig_dict to include the fused modules mapping to the qconfigs used for the modules that make up the fused modules. If the modules are not mapped to the same qconfig, then we will raise an error.

Test Plan: `python test/test_quantization.py TestFuseFx.test_qconfig_fused_module`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx][fix] Fused modules with object_type in qconfig"


Summary: When we do fusion, we replace certain modules (such as Linear + ReLU) with fused versions (such as LinearReLU) by calling `_fuse_fx` in prepare_fx. However when we try to look up using the fused module type in qconfig_dict, we cannot find a match anymore since the qconfig dict contains the original module types. An example is here [N882873](https://fburl.com/anp/azenjx3v). 

So we will now update the qconfig_dict to include the fused modules mapping to the qconfigs used for the modules that make up the fused modules. If the modules are not mapped to the same qconfig, then we will raise an error.

Test Plan: `python test/test_quantization.py TestFuseFx.test_qconfig_fused_module`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx][fix] Fused modules with object_type in qconfig"


Summary: When we do fusion, we replace certain modules (such as Linear + ReLU) with fused versions (such as LinearReLU) by calling `_fuse_fx` in prepare_fx. However when we try to look up using the fused module type in qconfig_dict, we cannot find a match anymore since the qconfig dict contains the original module types. An example is here [N882873](https://fburl.com/anp/azenjx3v). 

So we will now update the qconfig_dict to include the fused modules mapping to the qconfigs used for the modules that make up the fused modules. If the modules are not mapped to the same qconfig, then we will raise an error.

Test Plan: `python test/test_quantization.py TestFuseFx.test_qconfig_fused_module`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29406941).

supriyar commented on torch/quantization/fx/prepare.py:
```diff
@@ -166,6 +168,40 @@ def update_qconfig_for_qat(
+                    # different qconfigs specified in the qconfig_dict
+                    for op in ops:
+                        if object_type_dict.get(op, None) != fused_qconfig:
+                            raise LookupError(f"Need to specify the same qconfigs for both modules in {module_type}.")
```
Maybe extend the comment to say this happens during fusion

supriyar commented on torch/quantization/fx/prepare.py:
```diff
@@ -166,6 +168,40 @@ def update_qconfig_for_qat(
             object_type_dict[all_qat_mappings[k]] = v
     return qconfig_dict
 
+def update_qconfig_for_fusion(
```
Should we update the qconfig in the `fuse_fx` function?

angelayi replied:
It's a little easier to update in the current location because at that point the qconfig_dict and object_type has been converted to an ordered dictionary. But we could also update in the fuse_fx function, but I think the logic will be a little more complicated because we cannot just use `object_type_dict.get`.

angelayi committed:
Update on "[quant][fx][fix] Fused modules with object_type in qconfig"


Summary: When we do fusion, we replace certain modules (such as Linear + ReLU) with fused versions (such as LinearReLU) by calling `_fuse_fx` in prepare_fx. However when we try to look up using the fused module type in qconfig_dict, we cannot find a match anymore since the qconfig dict contains the original module types. An example is here [N882873](https://fburl.com/anp/azenjx3v). 

So we will now update the qconfig_dict to include the fused modules mapping to the qconfigs used for the modules that make up the fused modules. If the modules are not mapped to the same qconfig, then we will raise an error.

Test Plan: `python test/test_quantization.py TestFuseFx.test_qconfig_fused_module`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29406941](https://our.internmc.facebook.com/intern/diff/D29406941)

[ghstack-poisoned]

angelayi committed:
Update on "[quant][fx][fix] Fused modules with object_type in qconfig"


Summary: When we do fusion, we replace certain modules (such as Linear + ReLU) with fused versions (such as LinearReLU) by calling `_fuse_fx` in prepare_fx. However when we try to look up using the fused module type in qconfig_dict, we cannot find a match anymore since the qconfig dict contains the original module types. An example is here [N882873](https://fburl.com/anp/azenjx3v). 

So we will now update the qconfig_dict to include the fused modules mapping to the qconfigs used for the modules that make up the fused modules. If the modules are not mapped to the same qconfig, then we will raise an error.

Test Plan: `python test/test_quantization.py TestFuseFx.test_qconfig_fused_module`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29406941](https://our.internmc.facebook.com/intern/diff/D29406941)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29406941).
